### PR TITLE
First release of the framework

### PR DESCRIPTION
**Pull request by DeborahVolpe**
Pull request for obtaining a first mqt-toolkit-compatible release of the framework, adding documentation, managing libraries dependencies, updating code for supporting qiskit 1.0.

### CODE REVIEW

**Comment by burgholzer** [comment]
I pushed a couple of changes here that should help to stir this in the right direction.
First and foremost, I enabled Python 3.8 to 3.12 support. It should definitely be possible to support all these versions here.
Then, I cleaned up the dependency mess a little bit and tried to identify suitable versions.
I fixed whatever I could. However, I also left some things slightly broken and would ask you to please fix them. Most importantly, I updated all depenencies and some of the code to Qiskit 1.0. This broke some parts with QAOA, VQE, etc. There should be migration guides for all of these. So it is just a matter of time to fix the resulting errors.

The bigger thing, that is still unaddressed is concerning the warnings about invalid escape sequences in third party libraries. As far as I was able to tell, these concern DWave's libraries `dimod` + `dwave-samplers` and `docplex`. The main reason for this failure is that, since Python 3.11, such invalid escape sequences are no longer warnings, but SyntaxErrors. So this is really a fundamental flaw in these libraries. The appropriate thing to do is to raise issue in their repositories reporting the errors based on a minimal reproducible example. If a codebase is using ruff as a linter, there is even a check that prevents such sequences, see https://docs.astral.sh/ruff/rules/invalid-escape-sequence/. In most cases, you could probably fix those issues yourself and submit a PR to their repository. We would then wait for the respective PRs to be merged and a new release to be created. Then, we can use that release as a lowerbound here.

I also enabled the pre-commit bot for this repositore to reveal the respective warnings in PRs.

File: docs/Quickstart.ipynb
Comment (burgholzer) [inline docs/Quickstart.ipynb:1]: This is just copied from the predictor repo..
this should show an example of the QAO library.

**Review by burgholzer — CHANGES_REQUESTED** [review]
just a quick pass over the pr with a couple of suggestions. i will just quickly batch some of them up and commit them. then ill check the CI failures.

thanks for all your work on this

**Comment by burgholzer** [comment]
Ok. This is looking quite good already. All Ubuntu and Windows failures are due to the invalid escape sequences, but work otherwise. It also works on x86 (intel) macOS.
However, it does not work on Apple Silicon arm64 macOS runners. There it fails to build one of the dependencies (`rectangle-packer`), which does not provide wheels for arm64 macOS at the moment. It would be best to check their repository and its issue tracker. If not already, consider raising an issue there to provide the respective wheels.

please also always post links to relevant PRs and issues from other repositories over here once you created them.

File: noxfile.py
Comment (burgholzer) [inline noxfile.py:23]:
```suggestion
    "wheel>=0.40",
    "cython>=3.0",
```
This could actually fix the macOS issues.

File: .pre-commit-config.yaml
Comment (burgholzer) [inline .pre-commit-config.yaml:61]:
```suggestion
    rev: v0.4.7
```
new ruff update from #2

File: CITATION.bib
Comment (burgholzer) [inline CITATION.bib:1]: This is not a proper proceedings reference. Published work should not be under `@MISC`.
Please adjust the reference accordingly.
Also: the year should not be part of the conference name.

File: README.md
Comment (burgholzer) [inline README.md:20]:
```suggestion
MQT Quantum Auto Optimizer is a framework that allows one to automatically translate an optimization problem into a quantum-compliant formulation and to solve it with one of the main quantum solvers (Quantum Annealer, Quantum Approximate Optimization Algorithm, Variational Quantum Eigensolver and Grover Adaptive Search)
```

File: docs/Constraints.rst
Comment (burgholzer) [inline docs/Constraints.rst:19]: Please make sure that proper RST headlines are used.
The underlines must be as long as the header in all cases.
Sphinx should tell you that and scream at you when you try to build the documentation (`nox -s docs`)

When I try to build the docs, the following errors and warnings pop up:
```bash
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Constraints.rst:19: WARNING: Title underline too short.

Constraints declarations
----------------------
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Constraints.rst:19: WARNING: Title underline too short.

Constraints declarations
----------------------
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Problem.rst:20: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Problem.rst:28: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solution.rst:42: WARNING: Bullet list ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:38: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:51: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:76: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:94: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:111: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:121: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:141: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:154: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:173: WARNING: Block quote ends without a blank line; unexpected unindent.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst:184: ERROR: Unexpected indentation.
/home/lburgholzer/PycharmProjects/mqt-qao/docs/index.rst:26: WARNING: toctree contains reference to nonexisting document 'Variables'
/home/lburgholzer/PycharmProjects/mqt-qao/docs/index.rst:26: WARNING: toctree contains reference to nonexisting document 'Solvers'
looking for now-outdated files... none found
pickling environment... done
checking consistency... /home/lburgholzer/PycharmProjects/mqt-qao/docs/Solver.rst: WARNING: document isn't included in any toctree
/home/lburgholzer/PycharmProjects/mqt-qao/docs/Variable.rst: WARNING: document isn't included in any toctree
```

Please fix those.

File: docs/Problem.rst
Comment (burgholzer) [inline docs/Problem.rst:28]: For a first version, this is fine. But there are automated ways to generate API documentation from the package and its docstrings that should be preferred of manually writing this out, which is error prone and hard to keep in sync with the actual code.

File: docs/Solution.rst
Comment (burgholzer) [inline docs/Solution.rst:90]: Given that you added examples to all these sections (which is nice), it would be nice to turn these into actual examples that are run when the documentation is being built so that you can make sure that everything works smoothly. This doesn't have to happen in this PR though. But please create an issue for reworking/extending the documentation (can be combined with the comment on the automatic API documentation).

For inspiration, you can have a look at the documentation over at the MQT Core project: https://github.com/cda-tum/mqt-core/tree/main/docs

File: docs/refs.bib
Comment (burgholzer) [inline docs/refs.bib:1]: Please apply similar corrections here as for the CITATION.bib file

File: scripts/Knapsack/Knapsack.py
Comment (burgholzer) [inline scripts/Knapsack/Knapsack.py:1]: GitHub CodeQL points to a list of problems here. Please fix these.

File: scripts/linear_regression/LinearRegression.py
Comment (burgholzer) [inline scripts/linear_regression/LinearRegression.py:1]: GitHub CodeQL points to a list of problems here. Please fix these.

**Review by burgholzer — CHANGES_REQUESTED** [review]
This is very close to the finish line. Just a couple further fixes necessary before it can be merged. Please see the inline comments.

**Comment by burgholzer** [comment]
Also: Please provide a proper Title, a Description, and the appropriate Labels for this Pull Request.

File: CITATION.bib
Comment (DeborahVolpe) [reply CITATION.bib:1]: Done

File: docs/Constraints.rst
Comment (DeborahVolpe) [reply docs/Constraints.rst]: Done

File: scripts/Knapsack/Knapsack.py
Comment (DeborahVolpe) [reply scripts/Knapsack/Knapsack.py:1]: It should be ok now

File: scripts/linear_regression/LinearRegression.py
Comment (DeborahVolpe) [reply scripts/linear_regression/LinearRegression.py:1]: It should be ok now

**Review by burgholzer — APPROVED** [review]
Many thanks for all the work 🙏🏼 
Overall, this is looking good now and ready to merge ✅ 

However, https://github.com/cda-tum/mqt-qao/pull/1#issuecomment-2151856687 is still left unaddressed. Once that is done and CI has finished, I'll proceed to merge and create a release.

**Pull request merged**
